src/proxy/pool.rs

```diff
@@ -12,80 +12,539 @@
+            Ok(_guard) => {
+                // BEGIN take inner writelock
+                // If we get here, it means the following are true:
+                // 1. We did not get a connection for our key.
+                // 2. We have the exclusive inner writelock to create a new connection for our key.
+                //
+                // So, carry on doing that.
+                debug!("nothing else is creating a conn and we won the lock, make one");
+                let pool_conn = self.spawn_new_pool_conn(workload_key.clone()).await;
+                let client = ConnClient {
+                    sender: pool_conn?,
+                    stream_count: Arc::new(AtomicU16::new(0)),
+                    stream_count_max: self.max_streamcount,
+                    wl_key: workload_key.clone(),
+                };
+
+                debug!(
+                    "starting new conn for key {:#?} with pk {:#?}",
+                    workload_key, pool_key
+                );
+                Some(client)
+                // END take inner writelock
+            }
+            Err(_) => {
+                debug!("we didnt' win the lock, something else is creating a conn, wait for it");
+                // If we get here, it means the following are true:
+                // 1. At one point, there was a preexisting conn in the pool for this key.
+                // 2. When we checked, we got nothing for that key.
+                // 3. We could not get the exclusive inner writelock to add a new one for this key.
+                // 4. Someone else got the exclusive inner writelock, and is adding a new one for this key.
+                //
+                // So, loop and wait for the pool_watcher to tell us a new conn was enpooled,
+                // so we can pull it out and check it.
+                loop {
+                    match self.pool_watcher.changed().await {
+                        Ok(_) => {
+                            trace!(
+                                "notified a new conn was enpooled, checking for hash {:#?}",
+                                hash_key
+                            );
+                            // The sharded mutex for this connkey is already locked - someone else must be making a conn
+                            // if they are, try to wait for it, but bail if we find one and it's got a maxed streamcount.
+                            let existing_conn = self.state.connected_pool.get(&hash_key);
+                            match existing_conn {
+                                None => {
+                                    trace!("woke up on pool notification, but didn't find a conn for {:#?} yet", hash_key);
+                                    continue;
+                                }
+                                Some(e_conn) => {
+                                    debug!("found existing conn after waiting");
+                                    // We found a conn, but it's already maxed out.
```

Comment:
I am not sure this is what we want. Consider max steams=3 (for simple example).

Here is the desired behavior I think:
1. I open 2 streams. One wins, and makes a connection and opens a stream. The other waits, and opens a stream on the connection.
1. I open 2 more streams. One checks out the existing connection, _and the other opens a new connection_ .
1. (now I have 1 conn w/ 3, 1 conn w/ 1 stream)
2. Now I open 6 more streams. I should have 2 checkout the existing connection, and at the same time _open 2 more connections concurrently_

IIUC here, we are saying "You can never have more than `max_streamcount` _at all_". But that is what we are doing today. What we want is "you can never have more than max_streamcount on a connection (because that is how HTTP2 works), so if you go over we spin up more connections).

Reply:
Upon rereading, I misunderstood (which I was pretty sure going into it :slightly_smiling_face: ).

I do think on (4) we are not doing the new connections concurrently but

Reply:
> What we want is "you can never have more than max_streamcount on a connection (because that is how HTTP2 works), so if you go over we spin up more connections).

yes, that is what we have here.

That being said, we only check the max streamcount _on conn checkout attempts_ - so we cannot stop someone from checking out 1x and sending 300 streams over a conn we configured with `max_streamcount = 250`

Our code doesn't do that, and I did not want to update atomics on stream send as that's a perf hit, and we really don't need that granularity here.

Reply:
Yeah that's fine, we can just say "1 checkout == 1 stream" and trust we follow that.

But incrementing 1 atomic takes ~1ns. We don't need to worry about that in the context of something run once per connection (O(milliseconds))

Reply:
> But incrementing 1 atomic takes ~1ns. We don't need to worry about that in the context of something run once per connection (O(milliseconds))

Yep, that's fair - I wasn't sure how long it would take (and it seems to be platform-dependent too) and I don't strictly need to do it here, so I didn't.